src/Network/Transport/TCP.hs

```diff
@@ -629,6 +644,64 @@ apiNewEndPoint transport =
     resolveMulticastGroupError =
       TransportError ResolveMulticastGroupUnsupported "Multicast not supported"
 
+-- | Abstraction of a queue for an 'EndPoint'.
+--
+--   This specifies which 'Event's will come from
+--   'receive :: EndPoint -> IO Event' and when. It is highly general so that
+--   the simple yet potentially very fast implementation backed by a single
+--   unbounded channel can be used, without excluding more nuanced policies
+--   like class-based queueing with bounded buffers for each peer, which may be
+--   faster in certain conditions but probably has lower maximal throughput.
+--
+--   A 'QDisc' must satisfy some properties in order for the semantics of
+--   network-transport to hold true. To grok these, it may help to know how
+--   network-transport-tcp will use the 'QDisc'.
```

Comment:
I'm not 100% of the properties that @avieth is thinking of, but the main one I can think of is that the QDisc cannot drop events entirely (would break protocols). That is everything supplied via `qdiscEnqueue` must eventually become available via `qdiscDequeue` (ie in a sequence of concurrent enqueues and dequeues). This implies a no-starvation property.

In the documentation here we can also explain by example, that the obvious possible implementations for a queuing discipline are an unbounded queue, a bounded queue with fair semantics on blocking (e.g. MVar's FIFO semantics), or some other bounded queue with non-FIFO choice such as weighted fair choice.